src/util/HoldTrack.h

```diff
@@ -0,0 +1,120 @@
+#include <atomic>
+#include <map>
+#include <memory>
+#include <mutex>
+#include <set>
+#include <string>
+#include <thread>
+
+// qserv headers
+#include "util/Issue.h"
+
+// This header declarations
+namespace lsst::qserv::util {
+
+/// This class is used to track where a program is waiting for long periods of time.
+/// It is meant to be used to locate blocks on communication or mutex for very long
+/// periods of time.
+/// For this class to work CheckKeySet() needs to be called on a regular basis, one
+/// call every 5 minutes may be adequate.
+/// Time precision is limited to milliseconds, which could be an issue in loops or
+/// recursion. This can be worked around by changing the note, but that situation
+/// is not where this class is meant to be used.
+class HoldTrack {
+public:
+    using Ptr = std::shared_ptr<HoldTrack>;
+
+    /// Keys sorted by thread::id, sequence_number, uint64_t(time in milliseconds).
+    /// sequence number is needed to keep items in order as too many items may occur in 1ms.
+    using KeyType = std::pair<std::thread::id, std::pair<uint64_t, uint64_t>>;
+
+    HoldTrack() = delete;
+    HoldTrack(HoldTrack const&) = delete;
+    HoldTrack& operator=(HoldTrack const&) = delete;
+
+    ~HoldTrack() = default;
+
+    /// Setup the global instance for storing hold information.
+    /// Items that have existed for longer than `durationLimit` will be logged.
+    static void setup(double durationLimitSeconds);
+
+    /// Make a KeyType object given the current information.
+    static KeyType makeKey();
+
+    /// Return a string detailing keys that have existed longer than `durationLimit`.
+    static std::string CheckKeySet();
+
+    /// RAII helper class for HoldTrack.
+    class Mark {
+    public:
+        using Ptr = std::shared_ptr<Mark>;
+        Mark(Issue::Context const& ctx, std::string const& note) {
+            if (_enabled && _globalInstance != nullptr) {
+                _key = makeKey();
+                _globalInstance->_addKey(_key, ctx, note);
+                _inserted = true;
+            }
+        }
+
+        ~Mark() {
+            if (_inserted && _globalInstance != nullptr) {
+                _globalInstance->_removeKey(_key);
+            }
+        }
+
+    private:
+        bool _inserted = false;  ///< set to true when a mark is inserted into the set.
+        KeyType _key;            ///< that matches this Mark in the set.
+    };
+
+private:
+    HoldTrack(double durationLimitSeconds) : _durationLimitMillisec(durationLimitSeconds * 1000.0) {
+        _enabled = true;
+    }
+
+    /// Add `key` to _keyMap with ctx prepended to note.
+    void _addKey(KeyType const& key, Issue::Context const& ctx, std::string const& note);
+
+    /// Remove `key` from _keyMap.
+    void _removeKey(KeyType const& key);
+
+    /// Return the next sequence number.
+    static uint64_t _getSeq() { return _seq++; }
+
+    static std::atomic<bool> _enabled;  ///< Set to true to enable tracking.
```

Comment:
This attribute's initialization and use are not consistent. It's initialized to be `false` during static initialization of the process. It's set to be `true` when an instance of the "global" class `HoldTrack` gets created and this private variable is used outside of the class `HoldTrack` by another class `Mark` (even though, it's "internal").